### PR TITLE
ARC: Add support for ARC HS family on nSIM

### PR DESCRIPTION
**Pull request by abrodkin**
Now with generalized nSIM platfrom it's easy to add support of another ARC core family support in Zephyr which we do here.

Just use `BOARD=nsim_hs` instead of `nsim_em`.
Note existing Zephyr SDK is perfectly capable of building binaries for ARC HS cores so we should be all set.

### CODE REVIEW

**Comment by abrodkin** [comment]
Pushed v2 with removal of unused `CONFIG_CPU_xxx`.

**Comment by abrodkin** [comment]
Pushed v3 with updated docs for nsIM board.

File: boards/arc/nsim/doc/index.rst
Comment (dbkinder) [inline boards/arc/nsim/doc/index.rst]:
```suggestion
There are three supported board sub-configurations:
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
one little grammar fix, otherwise docs LGTM

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks!

File: boards/arc/nsim/doc/index.rst
Comment (ruuddw) [inline boards/arc/nsim/doc/index.rst]: what does 'normal' mean in HS context? Guess MMU and MPU are not (yet) supported?
Maybe call it 'base' instead of normal.

**Review by ruuddw — CHANGES_REQUESTED** [review]
Few minor remarks inline, but please add a 'nsim_hs.yaml' in the boards/arc/nsim directory so sanitycheck can be used for testing (sanitycheck -p nsim_hs).
Btw, I added and tested this, result looks good:
132 of 134 tests passed with 0 warnings

File: arch/arc/Kconfig
Comment (abrodkin) [reply arch/arc/Kconfig]: My idea was if such not very straight-forward code was never really used up until now (see we mentioned even wrong file name, https://github.com/zephyrproject-rtos/zephyr/pull/16662/commits/2f7990c8b5a012cc0484ea8ce3ab40dbc721b148) it might be a better idea to use atomic ops provided by the GCC which are known to work (even though I don't know any directed stress test for GCC's atomics) as they are used pretty often in complex software packages like Qt (and many others). At least for starters. And then once we make sure those hand-written atomics do work OK we may switch ARC HS cores to use custom version by default. That way we'll be moving with smaller steps rather than adding many major changes in 1 merge.

What if I create a new issue for switch ARC HS to custom atomics? But I guess it makes sense to do it after HS addition is merged in the master branch.

File: boards/arc/nsim/doc/index.rst
Comment (abrodkin) [reply boards/arc/nsim/doc/index.rst]: Correct! It was just a copy-paste of EM's description.

So what if I say:
```
nsim_hs which includes base ARC HS features, i.e. without PMU & MMU
```
?

**Comment by abrodkin** [comment]
The following changes were done in the re-spin:
1. Added sanity check scenario for `nsim_hs` (run `./scripts/sanitycheck -p nsim_hs`)
2. Updated description of nSIM platform with ARC HS
3. Added ARC HS specific `-mmpy-option` to accommodate 54b229fc422d9dc8f57b765d891fb3d0bdbf4a9f

**Comment by abrodkin** [comment]
In the last re-spin the following changes were done:
 * Added support for unaligned access strictly required for ARC HS cores and optional for ARC EM w/o security option (enabled by default for nSIM with ARC EM, AKA `nsim_em`). This was required to pass full sanity check test set for `nsim_hs` w/o any failures.
 * To properly support enabled unaligned access fixed STATUS32 tinkering code

**Review by vonhust — APPROVED** [review]

**Comment by vonhust** [comment]
This PR only can be merged after PR #17420 is merged

**Comment by abrodkin** [comment]
@ruuddw could you please check if your comments were all addressed and mark corresponding conversation as solved so that this PR becomes ready for merge?

**Comment by vonhust** [comment]
@ruuddw, this PR  is next one to merge.  Then we can go to arc smp->HSDK

**Review by ruuddw — APPROVED** [review]

**Comment by abrodkin** [comment]
@andrewboie may we merge this one as a prerequisite for SMP for ARC (https://github.com/zephyrproject-rtos/zephyr/pull/17747)?

File: arch/arc/Kconfig
Comment (andrewboie) [reply arch/arc/Kconfig]: Honestly, it's better to use the compiler builtins if they are available.

Comment (abrodkin) [reply arch/arc/Kconfig]: @andrewboie well it's up to the user's philosophy. I mean in the Linux kernel and U-Boot bootloader we prefer to have as few external dependencies as possible thus trying to have everything we need from libgcc, string functions etc right in the kernel's source tree.

Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.

But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that as it's much easier and allows us to reduce amount of hand-crafted assembly (which might very well have subtle issues since we're just human beings).

Comment (andrewboie) [reply arch/arc/Kconfig]: > Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.

Not sure what you mean by this.
The semantics of Zephyr's atomic APIs should be the same across arches. Otherwise this would break arch-agnostic code that uses them (futexes, etc)

> But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that 

For all our other arches, we prefer builtins if they exist. Arches that use the C implementations or custom assembly, AFAIK, all do this because the builtins are not implemented or not working for some reason.

Comment (abrodkin) [reply arch/arc/Kconfig]: > > Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.
> 
> Not sure what you mean by this.
> The semantics of Zephyr's atomic APIs should be the same across arches. Otherwise this would break arch-agnostic code that uses them (futexes, etc)

Indeed. I meant compiler built-in atomics might have semantics which differ from what one may expect in his particular project. Moreover I guess different compilers (not sure if that's applicable to the Zephyr today as probably only GCC is supposed to be used, but what if at some point other compilers are adopted like Clang/LLVM, proprietary ones like IAR or Keil etc) and even compiler versions might have a bit different semantics of atomics.

But probably I'm exaggerating here.
 
> > But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that
> 
> For all our other arches, we prefer builtins if they exist. Arches that use the C implementations or custom assembly, AFAIK, all do this because the builtins are not implemented or not working for some reason.

That's a good move and as you may see I'm all for using built-ins as more robust and simple solution. Then do you suggest to get rid of custom atomic ops for ARC at all?

Comment (andrewboie) [reply arch/arc/Kconfig]: >  Then do you suggest to get rid of custom atomic ops for ARC at all?

If we can, yes.
My memory is hazy, but I think the reason why those ASM implementations exist is because GCC didn't have builtins for ARC when the ARC port was added to Zephyr, but I'd have to investigate more to be sure.

Comment (andrewboie) [reply arch/arc/Kconfig]: > Indeed. I meant compiler built-in atomics might have semantics which differ from what one may expect in his particular project. Moreover I guess different compilers (not sure if that's applicable to the Zephyr today as probably only GCC is supposed to be used, but what if at some point other compilers are adopted like Clang/LLVM, proprietary ones like IAR or Keil etc) and even compiler versions might have a bit different semantics of atomics.

Ah I see. Yeah if a compiler has different atomic semantics then there would need to be a separate set of wrappers to the Zephyr atomic APIs. I expect LLVM would work the same as GCC, but who knows with other compilers.

Comment (abrodkin) [reply arch/arc/Kconfig]: > 
> 
> > Then do you suggest to get rid of custom atomic ops for ARC at all?
> 
> If we can, yes.


Ok will do it shortly as otherwise we do need to have a good and thorough study of custom implementation which I'd like to escape :)

> My memory is hazy, but I think the reason why those ASM implementations exist is because GCC didn't have builtins for ARC when the ARC port was added to Zephyr, but I'd have to investigate more to be sure.

I took a look already and at least from this git repo I see ARC custom atomics were added in the very first commit 8ddf82cf70dc6f951ab477f325dee0efde3ec589. 

What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to https://stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.

Comment (andrewboie) [reply arch/arc/Kconfig]: > What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.

You get linker errors if the atomics aren't actually implemented for that arch.
I ran into this when I wrote the Nios II port.

Comment (abrodkin) [reply arch/arc/Kconfig]: > 
> 
> > What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.
> 
> You get linker errors if the atomics aren't actually implemented for that arch.
> I ran into this when I wrote the Nios II port.

Indeed! Atomic ops for ARC were only added in the end of 2015 (https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/f5599934783a78e3e89c69a88c6cc2fe741c7afa) and made it to arc-2016.03 release which was published on December 9-th 2016 (https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/releases/tag/arc-2016.09-release). which was much later than the first commit in Zephyr git repo :)

Mystery solved!

**Pull request merged**
